chore(docs): ACS API guidelines - Part 1: Naming convention (#7617)
stackrox/stackrox: .github/api_best_practices.md

> # ACS API Guidelines

Reviewer: Please mention kinds of APIs this document focuses because they are different. REST, gRPC, GraphQL, Kubernetes (CRDs).

Author: This document intends to provide guidelines for all APIs, which will come in separate PRs. I don't see the point of enlisting each category here, for that there is a table of contents.

Reviewer: My point is in similar direction as Van's comment about URLs. The following section `Naming Guidelines` talks about gRPC and proto naming. How does the reader know that? Well, some amount of experience will allow inferring that context from the content itself, but possessing that context is already high enough bar.

I think, this bar is set higher than necessary because we'd want less experienced engineers to also benefit from reading these guidelines. Anyone can read anything, the question here is that we make it harder to comprehend.

The second, related, part of my point is that the presented naming guidelines make little sense for Kubernetes APIs which are mostly resource objects (nouns) instead of actions (verbs). If someone wants to put guidelines on Kubernetes APIs naming, how does that person know if this document is fit for that?

The last part of my point is that by making it explicit which API kinds will be in the document, you'd set expectations for reviewers about what they'll find in the future iterations. You'd also be able to source info from reviewers experience about important differences if certain guidelines don't apply for all kinds.

My honest fear is that the guidelines are going to default to gRPC and the rest will be an afterthought, at best. It's fine to address gRPC APIs only but let's be explicit about that. The opposite of doing gRPC with few touches about REST, GraphQL and/or Kubernetes would leave the first impression of something thorough but makes bitter taste when applied in practice because the other kinds are substantially different in some aspects to be covered in a manner of deviations from/extensions of gRPC.

Author: > The following section Naming Guidelines talks about gRPC and proto naming. How does the reader know that?

Fair point. The individual section title can be expanded like `Method Name` -> `RPC Method Name` and likewise. I will do that. 

I want to mention the PRs will be much smaller chunks, therefore, we should refrain from asking for things that are not the PR, and rather focus on things that are part of the PR.

help developers avoid common mistakes and build more robust APIs. When in doubt, It is encouraged to seek resolution
in slack channel [#forum-acs-api-design](https://redhat-internal.slack.com/archives/C05MMG2PP8A).

## Table of Contents

- [Naming Guidelines](#naming-guidelines)
  - [Service Name](#service-name)
  - [Method Name](#method-name)
  - [Message Name](#message-name)

## Naming Guidelines

### Service Name
Service name **must** be unique and should use a noun that generally refers to a resource or product component 
e.g. `DeploymentService`, `ReportService`, `ComplianceService`. Intuitive and well-known short forms or 
abbreviations may be used in some cases (and could even be preferable) for succinctness 
e.g. `ReportConfigService`, `RbacService`. 
All RPC methods grouped into a single service must generally pertain to the primary resource of the service.

All service defined in versioned package `/proto/api/v2` **must not** import from 
`/proto/storage` and `/proto/internalapi` packages.
All new RPC methods defined in versioned package `/proto/api/v1` **must not** import from

> `/proto/storage` and `/proto/internalapi` packages.

Reviewer: This paragraph feels to not belong to Service Naming. Rather, this talks about proto files layout in repo and import rules.
I suggest moving it to a separate section, even if it's just declared as a stub. We'll be able to rename and move when the text is in a separate section, but we'll likely forget to move it after a couple iterations if this text stays here.

Reviewer: These are guidelines but these aren't **naming** guidelines.

Anyhow, I don't feel pedantic enough to fight for that so it can stay here.

Author: I feel it impacts everything below, so I will leave it as is for now.

### Method Name

> Methods (and URLs) should be named such that they provide insights into the functionality. Typically, the method 

Reviewer 2: This is the only mention of URLs in this doc.

I would like us to specify that URLs should use hyphens to separate words in the URL.

It makes it easier to parse the URL quickly, and also reduces the chance of typos for users who are using the API at the command line.

You can see this in the GitHub API here, with the `dependency-graph` part of the URL,
https://docs.github.com/en/rest/dependency-graph/dependency-review?apiVersion=2022-11-28#get-a-diff-of-the-dependencies-between-commits

It's also part of Google's URL style guide for human-readable URLs,
https://developers.google.com/search/docs/crawling-indexing/url-structure

Author: @vjwilson good catch. The section for URL naming will be separate. I missed removing this one reference to URLs.

name should follow the _VerbNoun_ convention. For example, `StartComplianceScan`, `RunComplianceScan`, 
and `GetComplianceScan` are not the same. `StartComplianceScan` must return without waiting for the compliance 
scan to complete. `RunComplianceScan` may or may not wait for a compliance scan; the method name is ambiguous 
and should be avoided if it starts a process and returns without waiting. `GetComplianceScan` should not run 
a compliance scan but only fetches a stored one.

| Verb   | Noun           | Method name         |
|--------|----------------|---------------------|

> | List   | Deployment     | `ListDeployments`   |

Author: Yes, it will be addressed in one of the upcoming PRs.

| Get                       | Deployment     | `GetDeployment`     | `GetDeploymentRequest`     | `GetDeploymentResponse`     |   
| Update                    | Deployment     | `UpdateDeployment`  | `UpdateDeploymentRequest`  | `UpdateDeploymentResponse`  |  
| Delete                    | Deployment     | `DeleteDeployment`  | `DeleteDeploymentRequest`  | `google.protobuf.Empty`     |     
| Get                       | ReportStatus   | `GetReportStatus`   | `GetReportStatusRequest`   | `GetReportStatusResponse`   |  
| Run                       | ComplianceScan | `RunComplianceScan` | `RunComplianceScanRequest` | `RunComplianceScanResponse` | 

Avoid prepositions as much as possible (e.g. “for”, “with”; `DeploymentWithProcessInfo`, `DeploymentWithImageScan`).
In case such a need arises, add a field to the request message and response message.

<table>
<tr>
<td><b>Instead of</b></td><td><b>Use</b></td>
</tr>
<tr>
<td>

`GetDeploymentWithImageScanRequest`

</td>
<td>

```
GetDeploymentRequest {
  bool with_image_scan;
}
```

</td>
</tr>
<tr>
<td>

`GetDeploymentWithImageScanResponse`

</td>
<td>

```

> GetDeploymentWithImageScanResponse

Reviewer: Is there no mistake to have `GetDeploymentWithImageScanResponse` second time here?

GetDeploymentResponse {
  Deployment deployment;
  Image image;

> }

Reviewer: Not exactly in scope here, rather an inspired question. Does it happen that we need a deployment coupled with different other objects in _different_ response types? What's the recommendation then, I suppose we shouldn't create a mega response type with multiple optional attributes, right? A completely made-up example:
```
GetDeplomentResponse {
  Deployment deployment;
  Image image; // Will be set by GetDeploymentImageScan
  NetworkPolicy network_policy; // Will be set by InferDeploymentNetworkPolicy
  ProcessBaseline process_baseline; // Will be set by GetDeploymentProcessBaseline
}
```

I suppose we'll create multiple Response types, one per operation. Can we reuse `GetDeploymentResponse` name if it looks tempting?

Author: Your suggestions are completely valid and I will add them to examples.

I have come across customers who demand mega responses. Now, the mega response that we both are talking about could be different, are we talking about every storage field in the response? Not necessarily, but that is a separate topic about decoupling storage and APIs.

# ACS API Guidelines
An overview of the set of standardized guidelines and practices to ensure that the APIs across different 
projects and teams adhere to a common structure and design. This consistency simplifies the process of understanding 
and maintaining APIs. The guidelines helps identify potential pitfalls and design choices that could lead to issues to 
help developers avoid common mistakes and build more robust APIs. When in doubt, It is encouraged to seek resolution
in slack channel [#forum-acs-api-design](https://redhat-internal.slack.com/archives/C05MMG2PP8A).

## Table of Contents

- [Naming Guidelines](#naming-guidelines)
  - [Service Name](#service-name)
  - [RPC Method Name](#rpc-method-name)
  - [RPC Message Name](#rpc-message-name)

## Naming Guidelines

> ### Service Name

Author: I feel the suggestion is leaking implementation details and we should refrain from doing so at higher level. Therefore, if some good engineer implements a `XYZService` interface then the argument fails.

Nonetheless, there exists a service of type `APIServiceWithCustomRoutes` registered for each custom route, so there exists a grpc endpoint.

Service name **must** be unique and should use a noun that generally refers to a resource or product component 
e.g. `DeploymentService`, `ReportService`, `ComplianceService`. Intuitive and well-known short forms or 
abbreviations may be used in some cases (and could even be preferable) for succinctness 
e.g. `ReportConfigService`, `RbacService`. 
All RPC methods grouped into a single service must generally pertain to the primary resource of the service.

All service defined in versioned package [/proto/api/v2](https://github.com/stackrox/stackrox/blob/master/proto/v2)
**must not** import from [/proto/storage](https://github.com/stackrox/stackrox/blob/master/proto/storage) and
[/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi) packages.
All new RPC methods defined in versioned package [/proto/api/v1](https://github.com/stackrox/stackrox/blob/master/proto/v1) **must not** import from
`/proto/storage` and `/proto/internalapi` packages.

> ### RPC Method Name

Reviewer: (*)

Caring about less experienced readers, I suggest to stick with `gRPC` instead of "generic" `RPC` which can create confusion.

```suggestion
### gRPC Method Name
```

Please also search-replace `RPC` with `gRPC` in the entire doc.

Author: fair enough

> # ACS API Guidelines

Author: I will address your comment in a follow-up PR

An overview of the set of standardized guidelines and practices to ensure that the APIs across different 
projects and teams adhere to a common structure and design. This consistency simplifies the process of understanding 

> and maintaining APIs. The guidelines helps identify potential pitfalls and design choices that could lead to issues to 

Reviewer 3: ```suggestion
and maintaining APIs. These guidelines help identify potential pitfalls and design choices that could lead to issues to 
```

help developers avoid common mistakes and build more robust APIs. When in doubt, It is encouraged to seek resolution
in slack channel [#forum-acs-api-design](https://redhat-internal.slack.com/archives/C05MMG2PP8A).

## Table of Contents

- [Naming Guidelines](#naming-guidelines)
  - [gRPC Service Name](#grpc-service-name)
  - [gRPC Method Name](#grpc-method-name)
  - [gRPC Message Name](#grpc-message-name)

## Naming Guidelines

### gRPC Service Name

> Service name **must** be unique and should use a noun that generally refers to a resource or product component 

Reviewer 3: ```suggestion
The service name **must** be unique and should use a noun that generally refers to a resource or product component 
```

> Service name **must** be unique and should use a noun that generally refers to a resource or product component 

Author: Sticking to a standard naming convention when implementing services is a reasonable ask.

e.g. `DeploymentService`, `ReportService`, `ComplianceService`. Intuitive and well-known short forms or 
abbreviations may be used in some cases (and could even be preferable) for succinctness 

> e.g. `ReportConfigService`, `RbacService`. 

Reviewer 3: Not sure if this is too specific, but should we define case conventions? e.g. Should it be `RbacService` or `RBACService`?  (Sorry if this is some implied gRPC convention already.)

Author: It seems too specific to me. We have to let engineers use their best judgment.

All gRPC methods grouped into a single service must generally pertain to the primary resource of the service.

> Note:

Reviewer 3: This doesn't seem to fit under the naming section.  Should it instead be its own section?  If there is going to be another PR about API versioning, perhaps this note would fit better in that PR.

> - All service defined in versioned package [/proto/api/v2](https://github.com/stackrox/stackrox/blob/master/proto/v2)

Reviewer 3: ```suggestion
- All services defined in versioned package [/proto/api/v2](https://github.com/stackrox/stackrox/blob/master/proto/v2)
```

Author: irrelevant because the note has been removed

**must not** import from [/proto/storage](https://github.com/stackrox/stackrox/blob/master/proto/storage) and
[/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi) packages.
- All new gRPC methods defined in versioned package [/proto/api/v1](https://github.com/stackrox/stackrox/blob/master/proto/v1) 
**must not** import from `/proto/storage` and `/proto/internalapi` packages.
- All services of the type `APIServiceWithCustomRoutes` **must not** import from `/proto/storage` and `/proto/internalapi` packages.

### gRPC Method Name
Methods should be named such that they provide insights into the functionality. Typically, the method 
name should follow the _VerbNoun_ convention. For example, `StartComplianceScan`, `RunComplianceScan`, 

> and `GetComplianceScan` are not the same. `StartComplianceScan` must return without waiting for the compliance 

Reviewer 3: Do we have a standard verb for a method that starts a task and _does_ wait for it to complete?

Author: I think that should be accomplished through the request option. For example, `RunComplianceScan` is ambiguous, however, if its request is:
```
RunComplianceScanRequest {
  bool wait_for_scan_completion
} 
```

then the communication is clear.

Reviewer 3: That looks reasonable to me.  Can that be integrated into this paragraph somehow? e.g. "the method name is ambiguous and should be avoided unless fields in the message help clarify the meaning."

scan to complete. `RunComplianceScan` may or may not wait for a compliance scan; the method name is ambiguous 
and should be avoided if it starts a process and returns without waiting. `GetComplianceScan` should not run 
a compliance scan but only fetches a stored one.

| Verb   | Noun           | Method name         |
|--------|----------------|---------------------|
| List   | Deployment     | `ListDeployments`   |
| Get    | Deployment     | `GetDeployment`     |
| Update | Deployment     | `UpdateDeployment`  |
| Delete | Deployment     | `DeleteDeployment`  |
| Notify | Violation      | `NotifyViolation`   |
| Run    | ComplianceScan | `RunComplianceScan` |

> It is **recommended** that the verbs be imperative instead of inquisitive. Generally, the noun should be the resource type. 

Reviewer 3: Do you intend to be consistent with the use of bolding to indicate the strength of the suggestion?  i.e. recommended, should, may, must.  I see various places where these words are not bolded.

In some cases, the noun portion could be composed of multiple nouns e.g. `GetVulnerabilityDeferralState`, `RunPolicyScan`.

| Inquisitive               | Imperative                      |
|---------------------------|---------------------------------|
| `IsRunComplete`           | `GetRunStatus`                  |
| `IsAdmin`                 | `GetUserRole`                   |
| `IsVulnerabilityDeferred` | `GetVulnerabilityDeferralState` |

The noun portion of methods that act on a single resource **must** be singular e.g. `GetDeployment`. Those methods that 
act on the collection of resources **must** be plural e.g. `ListDeployments`, `DeleteDeployments`. Avoid prepositions 
(e.g. for, by) in method names as much as possible. Typically, this can be addressed by using a distinct verb, 
adding a field to the request message, or restructuring _VerbNoun_.

<table>
<tr>
<td><b>Instead of</b></td><td><b>Use</b></td>
</tr>
<tr>
<td>

`GetBaselineGeneratedNetworkPolicyForDeployment`

</td>
<td>

```
GenerateDeploymentNetworkPolicy

GenerateDeploymentNetworkPolicyRequest {
  bool from_baseline;  
  bool from_network_flows;

> }

Reviewer 4: While I get what you try to illustrate here (don't litter the URL, use query or body instead), all aspects of examples in guidelines are important. I doubt we want to encourage use of bool fields, so I recommend (also note correct formatting and proper proto syntax):
```
message GenerateDeploymentNetworkPolicyRequest {
    enum Source {
        UNKNOWN       = 0;
        BASELINE      = 1;
        NETWORK_FLOWS = 2;
    }

    Source from = 1;
}
```

Author: I specifically did not choose the enum path because it has the limitation that it does not provide setting multiple options.  For example, a network policy generation can consider both baseline and current net pols. This is where graphQL has a clear advantage.

e.g. use `GetDeploymentResponse` response instead of `Deployment`. This allows augmenting the response with 
supplemental information in the future.

| Verb                      | Noun           | Method name         | Request message            | Response message            |
|---------------------------|----------------|---------------------|----------------------------|-----------------------------|
| List                      | Deployment     | `ListDeployments`   | `ListDeploymentRequest`    | `ListDeploymentResponse`    |
| Get                       | Deployment     | `GetDeployment`     | `GetDeploymentRequest`     | `GetDeploymentResponse`     |   
| Update                    | Deployment     | `UpdateDeployment`  | `UpdateDeploymentRequest`  | `UpdateDeploymentResponse`  |  
| Delete                    | Deployment     | `DeleteDeployment`  | `DeleteDeploymentRequest`  | `google.protobuf.Empty`     |     
| Get                       | ReportStatus   | `GetReportStatus`   | `GetReportStatusRequest`   | `GetReportStatusResponse`   |  
| Run                       | ComplianceScan | `RunComplianceScan` | `RunComplianceScanRequest` | `RunComplianceScanResponse` | 

> Avoid prepositions as much as possible (e.g. “for”, “with”; `DeploymentWithProcessInfo`, `DeploymentWithImageScan`).

Reviewer 3: Given the imperative above to have messages based on their corresponding service name, do we need to repeat the requirement to avoid prepositions?  It seems like this is accomplished if the requirement is followed for service naming conventions alone.

Author: It is good to repeat, it helps with retention :)

In case such a need arises, add a field to the request message and response message.

<table>
<tr>
<td><b>Instead of</b></td><td><b>Use</b></td>
</tr>
<tr>
<td>

`GetDeploymentWithImageScanRequest`

</td>
<td>

```
GetDeploymentRequest {
  bool with_image_scan;
}
```

</td>
</tr>
<tr>
<td>

`GetDeploymentWithImageScanResponse`

</td>
<td>

```
GetDeploymentResponse {
  Deployment deployment;
  Image image;
}

Or,

GetDeploymentImageScanResponse {
  Image image;
}
```

</td>
</tr>
<tr>
<td>

`RunPolicyScanForDeploymentRequest`

</td>
<td>

`RunDeploymentPolicyScanRequest`

</td>
</tr>
</table>

All fields in the message **must** be lowercase and underscore separated names. The JSON names for the fields are 
autogenerated by the proto compiler. By default, field names are converted to camel case notation.

| Proto field name           | JSON field name        |
|----------------------------|------------------------|
| `network_data_start_time`  | `networkDataStartTime` |
| `expiry_date`              | `expiryDate`           |

Be explicit about conveying the specific purpose of fields e.g. instead of `expires_on`

> use `expiry_date`(/`timestamp`), instead of `network_data_since` use `network_data_start_time`.

Reviewer 4: What is the motivation for this one? I actually think that the suggested names are longer and not really more informative and hence inferior.

Author: Clarity is preferred. For example, `network_data_since` does not convey the unit; it may or may not include time portion.
